### ISSUE TITLE
Automatic backward/forward porting?

### ISSUE THREAD

**Issue by Yajo** [OCA/oca-github-bot#55]
This was discussed long ago at https://github.com/OCA/maintainer-tools/issues/381 but it seems that today the part of version bumping is solved by the bot and... well, we have a bot; so it might be a good moment to reopen the discussion.

One of the worst parts about code maintenance here in OCA is the presence of several versions of the same code across different git branches, [and I know it won't change](https://github.com/OCA/maintainer-tools/issues/334), but today we have this bot, which is a good foundation to finally attack this problem without altering current workflow.

Odoo itself uses the same system, but they actually have a combination of bot and humans to perform the merge+forwardport loop sanely, and [a different product and policy to offer](https://github.com/odoo/odoo/pull/25371#issuecomment-401382654), which helps their maintenance.

It would be great if the bot could handle forward ports (backward would be awesome too). To avoid the dangers of automatic code ports, all the bot should do is:

1. Am I instructed to do the forward port? then:
1. Try to rebase this patch onto the next version branch. Some special machinery will be needed to ignore conflicts in version bumps. Did it rebase? then:
   1. Open a PR to that new version.
   1. Assign it to whoever instructed the bot to forwardport the original PR.
1. If it didn't rebase, just add a comment there in the original PR specifying the reasons for not forwardporting. Humans will have to handle this case.

The same for backward ports (that could be left for the future).

A cool thing to have in mind would be massive version migrations. This again can be left for the future, but imagine a scenario where Odoo 13.0 is released, then the 13.0 branches are opened, and then the bot does by itself the commit and code forward ports and opens a PR per module.

This would let humans focus in the corner cases, increasing speed of development and reducing repetitive maintenance tasks :muscle:, which often get undone due to their nature. :blush:

**Comment by pedrobaeza** [comment]
It's a good idea, but the most one is the automatic migration try for all the modules, as the other is now more simple: as the version bumping is now done by the bot, it's way easier to just cherry-pick the PR commit without conflicts and push directly for a qualified maintainer, or create the pull request. Any way, an instruction in a PR like:

/ocabot cherry-pick `branch`

is very feasible, that performs:

1. Create branch starting out of `branch`.
2. Cherry-pick the commits of the PR (without extra ones that OCA Bot merger does on another branch)
3. Create a pull request with this.

And then, the usual review or `/ocabot merge` instruction process will act.

For the other thing (mass migration), it looks like more like a tool for maintainer-tools than something for the bot IMO.

**Comment by Yajo** [comment]
So you mean that you'd prefer that the forwardport should be the default behavior instead?

I like it. :sunglasses:

**Comment by pedrobaeza** [comment]
No, I mean that you instruct the bot to make the forward-port/backport for you.

**Comment by sbidoul** [comment]
On which repo would we create the source branch for the back/forward port PR?

If we create it on the OCA repo, contributors that are not PSC members will not be able to work on it.

**Comment by Yajo** [comment]
I like the way that Odoo chose. The fw port bot pings reviewers, mergers and contributors from original PR. Then if something goes wrong with the port, he notifies them every now and then. The original contributor can:

1. Add code suggestions.
1. Open its own PR superseding the automatic one.
1. Open a PR against the fwport branch.

Maybe the bot should display instructions on how to do any of those.

I guess it's a matter of time until github allows adding permissions to specific PR branches.

**Comment by sbidoul** [comment]
I made myself this little [tool](https://gist.github.com/sbidoul/a865d293adad10fd7e4dd32d71fa43b8) to facilitate back/forward porting or PRs.

**Comment by Yajo** [comment]
Cool, although this should be a bot job IMHO.
